### PR TITLE
Add Import functionlity

### PR DESCRIPTION
**Pull request by priyabhatnagar25**
Wrangler Import Functionality Common Steps for Framework

Please review @Vipinofficial11

### CODE REVIEW

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 1]
```diff
@@ -382,10 +383,8 @@ public static void clickOnAddEntityButton() {
   }
 
   public static void fixAllIfPrompted() {
-    if (SeleniumDriver.getDriver().findElements(By.xpath(CdfConnectionLocators.fixAllButtonXpath)).size() != 0) {
-      WaitHelper.waitForElementToBeDisplayed(CdfConnectionLocators.fixAllButtonElement());
-      WebElement fixAllButton = (SeleniumDriver.getDriver().findElement(CdfConnectionLocators.locatorOfFixAll()));
-      fixAllButton.click();
+    clickIfDisplayed(CdfConnectionLocators.locatorOfFixAll()); {
+      clickIfDisplayed(CdfConnectionLocators.locatorOfFixAll(), ConstantsUtil.SMALL_TIMEOUT_SECONDS);
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: Why do we need to call the same method again?

Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 2]
```diff
@@ -27,10 +27,19 @@
...
 import org.openqa.selenium.StaleElementReferenceException;
 import org.openqa.selenium.WebElement;
 
 import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+
+import static io.cdap.e2e.utils.ElementHelper.clickIfDisplayed;
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: Remove static Imports

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 3]
```diff
@@ -55,7 +64,7 @@ public static void openWranglerConnectionsPage() throws IOException {
    * Click on the Browser connections button inside Plugin's properties page
    */
   public static void clickBrowseConnectionsButton() {
-    ElementHelper.clickOnElement(CdfConnectionLocators.browseConnectionsButton);
+    clickOnElement(CdfConnectionLocators.browseConnectionsButton);
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: Once you remove the method import from above revert this change and use ```ElementHelper.clickOnElement()``` in all the methods.

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 4]
```diff
@@ -119,21 +128,21 @@ public static void waitTillConnectionDataLoadingCompletes(long timeoutInSeconds)
...
-   *
+   * <p>
    * Based on the connection type - table with rows containing data (i.e.bucket, dataset, table, directory names etc.)
    * will be displayed.
    * To select data for the connection use this method with its name as parameter.
    * Example : For GCS pass actual bucket name as parameter.
    */
   private static void selectConnectionDataRow(String dataRow) {
-    waitTillConnectionDataLoadingCompletes (ConstantsUtil.DEFAULT_TIMEOUT_SECONDS);
+    waitTillConnectionDataLoadingCompletes(ConstantsUtil.DEFAULT_TIMEOUT_SECONDS);
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java:129]: Revert this change

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 5]
```diff
@@ -146,15 +155,14 @@ private static void selectConnectionDataRow(String dataRow) {
 
   /**
    * Select data for connection using name
-   *
+   * <p>
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java:149]: Why added a <p> tag ?

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 6]
```diff
@@ -330,29 +338,29 @@ public static void verifyConnectionIsNotPresent(String connectionType, String co
       CdfConnectionLocators.locatorOfConnection(connectionType, actualConnectionName),
       ConstantsUtil.DEFAULT_TIMEOUT_SECONDS);
     Assert.assertTrue("Connection " + connectionName + " should not be present"
-      , !ElementHelper.isElementDisplayed(CdfConnectionLocators.locatorOfConnection(connectionType,
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java:333]: revert this change.

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 7]
```diff
@@ -366,4 +374,27 @@ public static void verifyConnectionErrorMessage(String errorMessageLocation) {
     AssertionHelper.verifyElementDisplayed(CdfConnectionLocators.locateConnectionFooterError);
     AssertionHelper.verifyElementContainsText(CdfConnectionLocators.locateConnectionFooterError, expectedErrorMessage);
   }
+
+  /**
+   * Click on the Plus Green Button
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: As discussed in the morning call the methods should be in Studio Actions class, move these methods there and make sure there is no change in this class.

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 8]
```diff
@@ -366,4 +374,27 @@ public static void verifyConnectionErrorMessage(String errorMessageLocation) {
...
+   */
+  public static void clickOnAddEntityButton() {
+    clickOnElement(CdfConnectionLocators.plusGreenButton);
+  }
+
+  public static void fixAllIfPrompted() {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: fixAllIfPrompted() to ```clickOnFixAllButton()```

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 9]
```diff
@@ -366,4 +374,27 @@ public static void verifyConnectionErrorMessage(String errorMessageLocation) {
...
+    clickIfDisplayed(CdfConnectionLocators.locatorOfFixAll());
+  }
+
+  /**
+   * Import json Files
+   */
+  public static void importJsonFiles(String s) throws URISyntaxException {
+    WaitHelper.waitForElementToBeDisplayed(CdfConnectionLocators.importButton, 30);
+    WebElement chooseFile = (SeleniumDriver.getDriver().findElement(CdfConnectionLocators.importuploadButton()));
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: Webelement should be in locators file.

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java [hunk 10]
```diff
@@ -366,4 +374,27 @@ public static void verifyConnectionErrorMessage(String errorMessageLocation) {
...
+  }
+
+  public static void fixAllIfPrompted() {
+    clickIfDisplayed(CdfConnectionLocators.locatorOfFixAll());
+  }
+
+  /**
+   * Import json Files
+   */
+  public static void importJsonFiles(String s) throws URISyntaxException {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: change the name of ```s``` to something relevant.

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 11]
```diff
@@ -26,6 +26,7 @@
  * Represents CDF locators related to connection pages and properties
  */
 public class CdfConnectionLocators {
+  public static String fixAllButtonXpath = "//button[@data-cy='fix-all-btn']";
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: fixAllButtonXpath to ```fixAllButton```

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 12]
```diff
@@ -94,4 +95,23 @@ public static WebElement locateConnectionAction(String connectionType, String co
 
   @FindBy(how = How.XPATH, using = "//div[@data-cy='connection-test-failure']")
   public static WebElement locateConnectionFooterError;
+
+
+  @FindBy(how = How.XPATH, using = "//img[@id='resource-center-btn']")
+  public static WebElement plusGreenButton;
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Name it something relevant instead of plusGreenButton.

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 13]
```diff
@@ -94,4 +95,23 @@ public static WebElement locateConnectionAction(String connectionType, String co
...
+
+  public static By importuploadButton() {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: You have already created the import button locator below. What is this one doing ?

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 14]
```diff
@@ -94,4 +95,23 @@ public static WebElement locateConnectionAction(String connectionType, String co
...
+    return By.xpath("//input[@id='resource-center-import-pipeline']");
+  }
+
+  @FindBy(how = How.XPATH, using = "//*[@id='import-pipeline']")
+  public static WebElement importButton;
+
+  public static By locatorOfFixAll() {
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Instead of creating a string of ```public static String fixAllButtonXpath = "//button[@data-cy='fix-all-btn']"``` you can directly create a locator using xpath.

File: src/main/java/io/cdap/e2e/utils/SeleniumDriver.java [hunk 15]
```diff
@@ -49,7 +49,7 @@ public class SeleniumDriver {
     ChromeOptions chromeOptions = new ChromeOptions();
     chromeOptions.addArguments("--no-sandbox");
     chromeOptions.addArguments("--disable-setuid-sandbox");
-    chromeOptions.addArguments("--headless");
+//    chromeOptions.addArguments("--headless");
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/e2e/utils/SeleniumDriver.java]: Revert this change

File: src/main/java/stepsdesign/PipelineSteps.java [hunk 16]
```diff
@@ -659,4 +665,18 @@ public void clickOnCreateButton () {
...
+  public void clickOnPlusGreenButton () {
+    CdfConnectionActions.clickOnAddEntityButton();
+  }
+
+  @Then("Select the json files for importing the pipelines for the plugin {string}")
+  public void selectJSONFiles(String path) throws URISyntaxException {
+    CdfConnectionActions.importJsonFiles(PluginPropertyUtils.pluginProp(path));
+  }
+
+  @When("Click on the fix all button if displayed")
```
Comment (Vipinofficial11) [inline src/main/java/stepsdesign/PipelineSteps.java]: You are using the fixAllButton method when you are uploading the pipeline I don't think you need a step for this.

File: src/main/java/stepsdesign/PipelineSteps.java [hunk 17]
```diff
@@ -659,4 +665,18 @@ public void clickOnCreateButton () {
   public void verifyInvalidConnectionErrorMessage(String errorMessageLocation) {
     CdfConnectionActions.verifyConnectionErrorMessage(errorMessageLocation);
   }
+
+  @Then("Click on the Plus Green Button to import the pipelines")
+  public void clickOnPlusGreenButton () {
+    CdfConnectionActions.clickOnAddEntityButton();
+  }
+
+  @Then("Select the json files for importing the pipelines for the plugin {string}")
```
Comment (Vipinofficial11) [inline src/main/java/stepsdesign/PipelineSteps.java]: Discuss this with Poojan once instead of writing the step for Json file you can make this step generic by creating the Utility class and adding a method for file upload.

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 14 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: resolved

File: src/main/java/stepsdesign/PipelineSteps.java (hunk 16 above)
Comment (priyabhatnagar25) [reply src/main/java/stepsdesign/PipelineSteps.java]: resolved

File: src/main/java/stepsdesign/PipelineSteps.java (hunk 17 above)
Comment (priyabhatnagar25) [reply src/main/java/stepsdesign/PipelineSteps.java]: resolved

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 13 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: One is just for finding the element and the second locator is just used for uploading the json file

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 12 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: resolved

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 11 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 10 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 9 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 8 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 7 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 6 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java:333]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 5 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java:149]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 2 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java (hunk 3 above)
Comment (priyabhatnagar25) [reply src/main/java/io/cdap/e2e/pages/actions/CdfConnectionActions.java]: resolved

File: src/main/java/io/cdap/e2e/utils/FileImportUtil.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+
+/**
+ * File Import Util
+ */
+public class FileImportUtil {
+
+  public static void uploadFile(String filePath) throws URISyntaxException {
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/utils/FileImportUtil.java]: This is generic file upload utility. It should have a file upload button element as a parameter

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java [hunk 19]
```diff
@@ -265,4 +265,29 @@ public static By locatorOfLoadingSpinnerOnValidateButton() {
   @Deprecated
   @FindBy(how = How.XPATH, using = "//*[@data-cy='plugin-validation-error-msg']")
   public static WebElement pluginValidationErrorMsg;
+
+  @FindBy(how = How.XPATH, using = "//img[@id='resource-center-btn']")
+  public static WebElement uploadFileButton;
+
+  public static By importuploadButton() {
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Rename: importPipelineInputTag

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java [hunk 20]
```diff
@@ -265,4 +265,29 @@ public static By locatorOfLoadingSpinnerOnValidateButton() {
...
+    return By.xpath("//input[@id='resource-center-import-pipeline']");
+  }
+
+  @FindBy(how = How.XPATH, using = "//*[@id='import-pipeline']")
+  public static WebElement importFileButton;
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Rename: importPipelineButton

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 19 above)
Comment (poojantcs) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: This is not used outside Locators class. Remove this and use the locator in "chooseFile" WebElement

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java [hunk 21]
```diff
@@ -265,4 +265,29 @@ public static By locatorOfLoadingSpinnerOnValidateButton() {
...
+
+  public static By locatorOfFixAll() {
+    return By.xpath("//button[@data-cy='fix-all-btn']");
+  }
+
+  public static WebElement chooseFile() {
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Rename: importPipelineInputTag

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java [hunk 22]
```diff
@@ -265,4 +265,29 @@ public static By locatorOfLoadingSpinnerOnValidateButton() {
...
+    return SeleniumDriver.getDriver().findElement(CdfStudioLocators.importuploadButton());
+  }
+
+  @FindBy(how = How.XPATH, using = "//div[@data-cy='pipeline-metadata']")
+  public static WebElement appendPipeline;
+
+  public static By renamePipeline() {
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Rename: renamePipelineLocator

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java [hunk 23]
```diff
@@ -265,4 +265,29 @@ public static By locatorOfLoadingSpinnerOnValidateButton() {
...
+  public static By locatorOfFixAll() {
+    return By.xpath("//button[@data-cy='fix-all-btn']");
+  }
+
+  public static WebElement chooseFile() {
+    return SeleniumDriver.getDriver().findElement(CdfStudioLocators.importuploadButton());
+  }
+
+  @FindBy(how = How.XPATH, using = "//div[@data-cy='pipeline-metadata']")
+  public static WebElement appendPipeline;
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Locator already exists. Reuse: pipelineName

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 24]
```diff
@@ -27,6 +27,7 @@
  */
 public class CdfConnectionLocators {
 
+
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Remove unnecessary new line

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java [hunk 25]
```diff
@@ -94,4 +95,5 @@ public static WebElement locateConnectionAction(String connectionType, String co
 
   @FindBy(how = How.XPATH, using = "//div[@data-cy='connection-test-failure']")
   public static WebElement locateConnectionFooterError;
+
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Remove unnecessary new line

File: src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java [hunk 26]
```diff
@@ -498,11 +505,40 @@ public static void closeStatusBanner() {
...
+
+  public static void importPipeline(String filePath) throws URISyntaxException {
+    FileImportUtil.uploadFile(filePath);
+  }
+
+  public static void renameThePipeline() {
+    WaitHelper.waitForElementToBeOptionallyDisplayed(CdfStudioLocators.renamePipeline(), 100);
+    ElementHelper.clickOnElement(CdfStudioLocators.appendPipeline);
+    pipelineName = "TestPipeline-" + RandomStringUtils.randomAlphanumeric(10);
+    renamePipelineNameAndSave(pipelineName);
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java]: Method already exists. Reuse: fillPipelineNameAndSave

File: src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java [hunk 27]
```diff
@@ -498,11 +505,40 @@ public static void closeStatusBanner() {
...
+  }
+  public static void renamePipelineNameAndSave(String pipelineName) {
```
Comment (poojantcs) [inline src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java]: Remove this method. And reuse: fillPipelineNameAndSave to rename the pipeline

File: src/main/java/io/cdap/e2e/utils/FileImportUtil.java (hunk 18 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/utils/FileImportUtil.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 19 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 20 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 21 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 22 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java (hunk 23 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfStudioLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 24 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java (hunk 25 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/locators/CdfConnectionLocators.java]: Done.

File: src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java (hunk 26 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java]: Done

File: src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java (hunk 27 above)
Comment (bharatgulati) [reply src/main/java/io/cdap/e2e/pages/actions/CdfStudioActions.java]: Done.
